ams/server/src/main/java/com/netease/arctic/server/persistence/converter/Object2ByteArrayConvert.java

```diff
@@ -17,7 +17,7 @@ public class Object2ByteArrayConvert<T> implements TypeHandler<T> {
   @Override
   public void setParameter(PreparedStatement ps, int i, T parameter, JdbcType jdbcType) throws SQLException {
     if (parameter == null) {
-      ps.setNull(i, Types.BLOB);
+      ps.setNull(i, Types.BINARY);
```

Comment:
What is the reason for this modification?

Reply:
Postgres require `BINARY` for `bytea` column type

Reply:
```
### The error may exist in com/netease/arctic/server/persistence/mapper/OptimizingMapper.java (best guess)
### The error may involve com.netease.arctic.server.persistence.mapper.OptimizingMapper.insertTaskRuntimes-Inline
### The error occurred while setting parameters
### SQL: INSERT INTO task_runtime (process_id, task_id, retry_num, table_id, partition_data, start_time, end_time, status, fail_reason, optimizer_token, thread_id, rewrite_output, metrics_summary, properties) VALUES     (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)  ,  (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?,?)
### Cause: org.postgresql.util.PSQLException: ERROR: column "rewrite_output" is of type bytea but expression is of type oid
  Hint: You will need to rewrite or cast the expression.
  Position: 259
	at org.apache.ibatis.exceptions.ExceptionFactory.wrapException(ExceptionFactory.java:30) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.update(DefaultSqlSession.java:199) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.session.defaults.DefaultSqlSession.insert(DefaultSqlSession.java:184) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperMethod.execute(MapperMethod.java:62) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperProxy$PlainMethodInvoker.invoke(MapperProxy.java:152) ~[mybatis-3.5.6.jar:3.5.6]
	at org.apache.ibatis.binding.MapperProxy.invoke(MapperProxy.java:85) ~[mybatis-3.5.6.jar:3.5.6]
	at com.sun.proxy.$Proxy49.insertTaskRuntimes(Unknown Source) ~[?:?]
	at com.netease.arctic.server.optimizing.OptimizingQueue$TableOptimizingProcess.lambda$null$6(OptimizingQueue.java:571) ~[amoro-ams-server-0.5.0.jar:?]
	at com.netease.arctic.server.persistence.PersistentBase.doAs(PersistentBase.java:48) ~[amoro-ams-server-0.5.0.jar:?]
	... 26 more 
```

Reply:
I think not, I test it locally and work fine

Reply:
ok

Reply:
Let me check db type before set `NULL` value

Reply:
In JDBC, Types.BINARY and Types.BLOB are constants used to represent binary data types. Their meanings differ slightly in MySQL and PostgreSQL, as follows:

In MySQL:

Types.BINARY: represents the BINARY type in MySQL, which is used to store fixed-length binary data.
Types.BLOB: represents the BLOB type in MySQL, which is used to store variable-length binary data.

When using MySQL, the ResultSet.getBytes() method can be used to retrieve data of BINARY or BLOB type. For BINARY type, the length of the returned byte array is equal to the length specified when the field was defined. For BLOB type, the length of the returned byte array is equal to the actual length of the data stored.

In PostgreSQL:

Types.BINARY: represents the BYTEA type in PostgreSQL, which is used to store variable-length binary data.
Types.BLOB: has no corresponding type in PostgreSQL, so the Types.BLOB constant should not be used when working with PostgreSQL.

When using PostgreSQL, the ResultSet.getBytes() method can be used to retrieve data of BYTEA type. The length of the returned byte array is equal to the actual length of the data stored.

In summary, the meanings of Types.BINARY and Types.BLOB differ slightly in MySQL and PostgreSQL, and the appropriate choice should be made based on the specific database type and use case. It is important to distinguish and use them correctly when working with these databases.


ChatGPT gave me the above answer.